<a href="https://colab.research.google.com/github/Melzr/Numerico/blob/main/TP1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

2. Hallar $\pi$ por dos caminos

(a) Programar un algoritmo para aproximar $\pi$ utilizando la funcion $\sin{x}$ con el metodo de
Newton-Raphson, en funcion de $x$, que realice iteraciones hasta alcanzar el limite de la herramienta utilizada.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import math

#x = np.linspace(0, 2*np.pi, 100)
# F = np.sin(x)
# dF = np.cos(x)

def F(x):
  return math.sin(x)

def dF(x):
  return math.cos(x)

#recibe una semilla, la funcion y su derivada
def newton_raphson(x0, f, df):
  
  xn = (x0 - f(x0)/df(x0))
  xa = x0
  error = abs(xn - xa)

  #error es 0 cuando se alcanza el limite de la representacion en python
  while (error > 0): 
    print(xn)
    xa = xn
    xn = (xa - f(xa)/df(xa))
    error = abs(xn - xa)

  return xn


#plt.figure(1)
#plt.plot(x, F)
#plt.show()

raiz = newton_raphson(3, F, dF)
print(raiz)

3.142546543074278
3.141592653300477
3.141592653589793
3.141592653589793


(b) Programar un algoritmo para aproximar $\pi$ utilizando la serie de Leibniz, en funcion de $n$.

In [ ]:
def leibniz(n):
  suma = 0
  for i in range(0, n):
    suma += ((-1)**i)/(2*i + 1)

  return suma*4

3.1415826535897198

(d) Ejecutar los programas solicitados en a y b utilizando representacion de punto 
otante de 64 bits y comparar las respuestas obtenidas con $n = 10$, $n = 100$, $n = 1000$, $n = 10000$ y $n = 100000$.

In [ ]:
print(leibniz(10))
print(leibniz(100))
print(leibniz(1000))
print(leibniz(10000))
print(leibniz(100000))


3.0418396189294032
3.1315929035585537
3.140592653839794
3.1414926535900345
3.1415826535897198


(e) Ejecutar los programas solicitados en a y b con una calculadora (aclarar marca y modelo) y
comparar las respuestas obtenidas con $n = 10$, $n = 100$, $n = 1000$, $n = 10000$ y $n = 100000$
(en caso de no alcanzar la memoria de la calculadora utilizar el maximo n posible).

Modelo `casio fx-570LA X`

Leibniz:

  * con $n=10$:  $3,232315809$

  * con $n=100$:  $3,151493401$

  * con $n=1000$:  $3,142591654$

  * con $n=10000$:  $3,141692644$  

(f) Representar las dos respuestas finales obtenidas (para n= 100000 y el ḿetodo de NewtonRaphson) en c, d y e de manera de expresarlo como $\pi= \bar{\pi} + ∆\pi $ 

Con el metodo Newton-Raphson: $\pi= 3.141592653589793 \pm 0.000000000000001$

Con la serie de Leibniz: $\pi= 3.1415826535897198 \pm 0.0000000000000001$


(g)  Podemos afirmar qué para la computadora el número π es una constante?


In [ ]:
np.float32(leibniz(10))

3.0418396